File: Contents/code/python/chapter2/lambdas.py
from typing import Callable

# A standard named function
def add_one_named(x: int) -> int:
    return x + 1

# The Lambda Calculus equivalent: an anonymous function
# Syntax: lambda bound_variable: body
add_one_lambda: Callable[[int], int] = lambda x: x + 1

# Beta-reduction in Python (Application)
# (lambda x: x + 1)(5)
result = (lambda x: x + 1)(5)
print(result) # Output: 6

File: Contents/code/haskel/chapter2/lambdas.hs
-- A standard named function
addOneNamed :: Int -> Int
addOneNamed x = x + 1

-- The Lambda Calculus equivalent: an anonymous function
-- Syntax: \bound_variable -> body
-- (The backslash '\' looks like a lambda 'lambda')
addOneLambda :: Int -> Int
addOneLambda = \x -> x + 1

-- Beta-reduction in Haskell (Application)
-- (\x -> x + 1) 5
result :: Int
result = (\x -> x + 1) 5
-- result evaluates to 6
